### PR TITLE
feat(wallet-api): device.transport LLD & LLM integration [LIVE-4293]

### PR DESCRIPTION
**Pull request by Justkant**
### 📝 Description

Expose a `hw-transport` using 3 RPC calls `device.getTransport`, `device.exchange` and `device.close` on the client & server.
Create a specific DeviceAction using a subject to maintain a single device transport open in the wallet.

### ❓ Context

- **Impacted projects**: `ledger-live-desktop`, `ledger-live-mobile`, `ledger-live-common` 
- **Linked resource(s)**: [`LIVE-4293`](https://ledgerhq.atlassian.net/browse/LIVE-4293) 

### ✅ Checklist

- [ ] **Test coverage** Manual testing using the dummy-wallet-app 
- [x] **Atomic delivery** 
- [x] **No breaking changes** 

### 📸 Demo

https://user-images.githubusercontent.com/3009753/208748293-43f6f8c1-54e5-4abd-8ec7-b0ea140bdfea.mov

#### Manifest example
```json
{
  "id": "dummy",
  "name": "Dummy Wallet App",
  "url": "http://localhost:3000",
  "homepageUrl": "http://localhost:3000",
  "icon": "",
  "platform": "all",
  "apiVersion": "^2.0.0 || ~0.0.1",
  "manifestVersion": "1",
  "branch": "experimental",
  "categories": [],
  "currencies": ["*"],
  "content": {
    "shortDescription": {
      "en": ""
    },
    "description": {
      "en": ""
    }
  },
  "permissions": [
    "account.list",
    "account.receive",
    "account.request",
    "currency.list",
    "device.close",
    "device.exchange",
    "device.transport",
    "message.sign",
    "transaction.sign",
    "transaction.signAndBroadcast",
    "wallet.capabilities"
  ],
  "domains": ["http://*", "https://*"]
}
```

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

File: libs/ledger-live-common/src/hw/getTransport.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                data: response.toString("hex"),
+              })
+            )
+            .catch((error) => subject.next({ type: "error", error }));
+        }
+      },
+    });
+
+    return subject;
+  }).toPromise();
```
Comment (chabroA) [inline libs/ledger-live-common/src/hw/getTransport.ts:57]: Why converting the `job` function to a promise?
The `job` function returns an `Observable`, not a promise 🤔

```ts
const withDevice: (deviceId: string) => <T>(job: (t: Transport) => Observable<T>) => Observable<T>
```

cf. https://github.com/LedgerHQ/ledger-live/blob/e58610f63dfa3ae1d213741a063390971a7e1587/libs/ledger-live-common/src/hw/deviceAccess.ts#L83

Since subject is an `Observable`, shouldn't it be like that instead:

```ts
  withDevice(deviceId)((transport) => {
    // The input part of the bidirectional communication
    subject.subscribe({
      ...
    })

    return subject;
  });
```

File: apps/ledger-live-mobile/src/components/WebPlatformPlayer/WebViewV2.tsx [hunk 2]
```diff
@@ -412,6 +421,97 @@ export const WebView = ({ manifest, inputs }: Props) => {
...
+                deviceId,
+              });
+
+              // Clean the ref on completion
+              transport.current.subscribe({
+                complete: () => {
+                  transport.current = undefined;
+                },
+              });
+              resolve("1");
```
Comment (chabroA) [inline apps/ledger-live-mobile/src/components/WebPlatformPlayer/WebViewV2.tsx:453]: Why `"1"`?

File: libs/ledger-live-common/src/hw/getTransport.ts (hunk 1 above)
Comment (Justkant) [reply libs/ledger-live-common/src/hw/getTransport.ts:57]: > Why converting the job function to a promise?

The call to `toPromise()` subscribes to the observable and calls the inner function, if we don't do that nothing happens.

The subject is then returned because it is actually the observable you receive from `withDevice` but we need the Subject interface.

File: apps/ledger-live-mobile/src/components/WebPlatformPlayer/WebViewV2.tsx (hunk 2 above)
Comment (Justkant) [reply apps/ledger-live-mobile/src/components/WebPlatformPlayer/WebViewV2.tsx:453]: This acts as the `transportId` for ledger live, we only use `"1"` here because ledger-live will only handle one transport at a time (this can change later one)

I still need to rename the `deviceId` to `transportId`

File: libs/ledger-live-common/src/hw/getTransport.ts (hunk 1 above)
Comment (gre) [reply libs/ledger-live-common/src/hw/getTransport.ts:57]: ideally , this should return an instance of `Transport` because what you call `Subject<BidirectionalEvent>` sounds equivalent to a Transport. (which also have a "disconnect" event system for errors, etc..)

saying this here because i'm actually going to need this soon for https://ledgerhq.atlassian.net/wiki/spaces/WALLETCO/pages/3948118370/tech+debt+documentation

Comment (gre) [reply libs/ledger-live-common/src/hw/getTransport.ts:57]: actually i see you need it for the "command" serialisation, i think i would just reword this to something else to avoid confusion.

maybe "openTransportAsSubject" or something.

ideally we'll then be able to "create a transport from a command" on the renderer side 🤔

Comment (Justkant) [reply libs/ledger-live-common/src/hw/getTransport.ts:57]: Understood, quick question @gre, is it fine for you if we do the changes incrementally after having this merged ?

File: apps/ledger-live-desktop/src/internal/commandHandler.js [hunk 3]
```diff
@@ -33,8 +45,10 @@ export function executeCommand(command: *, send: *) {
   const startTime = Date.now();
   logger.onCmd("cmd.START", id, 0, data);
   try {
-    subscriptions[requestId] = cmd(data).subscribe({
+    subscriptions[requestId] = cmd(data); // Store the subject returned instead of the subscription
+    subscriptions[requestId].subscribe({
       next: data => {
+        if (data?.type === "input-frame") return; // TODO Input can't be treated as output too
```
Comment (gre) [inline apps/ledger-live-desktop/src/internal/commandHandler.js:51]: not ideal to hack some knowledge of the commands in the generic rpc logic. let's figure out something

File: libs/ledger-live-common/src/hw/getTransport.ts (hunk 1 above)
Comment (gre) [reply libs/ledger-live-common/src/hw/getTransport.ts]: i think it's ok, but it could be interesting to confirm together this model will be future proof and compatible with the idea to rebuild a `Transport` instance on the renderer side :)

**Comment by gre** [comment]
can you make sure to document in the PR, at least by a mention (a schema might be even better), that this PR also evolve LLD commands internal process communication to make it bi-directional? It wasn't clear at first when I reviewed the code as it wasn't mentioned 👍

File: apps/ledger-live-desktop/src/internal/commandHandler.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (gre) [reply apps/ledger-live-desktop/src/internal/commandHandler.js:51]: yeah i deleted my msg after understanding this starts introduce a bidirectional com', worth mentioning in the main PR.

**Comment by gre** [comment]
Untitled-2021-05-24-1611

**Review by chabroA — APPROVED** [review]

File: apps/ledger-live-desktop/src/internal/transportHandler.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+          .exchange(Buffer.from(e.apduHex, "hex"))
+          .then(response =>
+            process.send?.({
+              type: transportExchangeChannel,
+              data: response.toString("hex"),
+              requestId: e.requestId,
+            }),
+          )
+          .catch(error =>
+            process.send?.({ type: transportExchangeChannel, error, requestId: e.requestId }),
```
Comment (gre) [inline apps/ledger-live-desktop/src/internal/transportHandler.ts:45]: you may want to serializeError to mirror the deserializeError (until we find a better way, but at least this way was proven working and was done in the "commands")

File: apps/ledger-live-desktop/src/internal/transportHandler.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+export const transportOpen = ({
+  data,
+  requestId,
+}: {
+  data: { descriptor: string };
+  requestId: string;
+}) => {
+  const subjectExist = transports.get(data.descriptor);
+  if (subjectExist) {
+    // Should we return saying the transport is open ?
```
Comment (gre) [inline apps/ledger-live-desktop/src/internal/transportHandler.ts:20]: probably not. it indicates a logic problem in our app.

I wonder therefore if there are a better `Error` to use so we can more clearly identify and remap it to a "worded error".

**Review by gre — COMMENTED** [review]
great rework!
I think there is possibly something to see to proxy the errors 'correctly' and make sure any instanceof will work like before.

serializeError and deserializeError can be used for that purpose (lib @ledgerhq/errors)

we also may need to consider specific Error for some of the `new Error` code 🤔

File: apps/ledger-live-desktop/src/internal/transportHandler.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    process.send?.({
+      type: transportOpenChannel,
+      error: new Error("Transport already opened for the given descriptor"),
+      requestId,
+    });
+    return;
+  }
+
+  withDevice(data.descriptor)(transport => {
```
Comment (gre) [inline apps/ledger-live-desktop/src/internal/transportHandler.ts:29]: you may not need to use withDevice. it's possible what yo just want is to `await` on the generic `open(` because it's only on RENDERER side that we will protect race conditions. you already do a protection with the "one transport at a time"

File: apps/ledger-live-desktop/src/internal/transportHandler.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+  requestId,
+}: {
+  data: { descriptor: string; apduHex: string };
+  requestId: string;
+}) => {
+  const subject = transports.get(data.descriptor);
+  if (!subject) {
+    process.send?.({
+      type: transportExchangeChannel,
+      error: new Error("No open transport for the given descriptor"),
```
[the remark this answers is not in the document]
Comment (Justkant) [reply apps/ledger-live-desktop/src/internal/transportHandler.ts]: done

**Comment by gre** [comment]
@Justkant 
```

/home/runner/work/ledger-live/ledger-live/apps/ledger-live-desktop/src/main/internal-lifecycle.js
  4:16  error  'uuidv4' is defined but never used  no-unused-vars
  ```

**Review by gre — APPROVED** [review]
clean! I like how you managed to simplify over the iterations 👍

File: apps/ledger-live-desktop/src/internal/transportHandler.ts (hunk 6 above)
Comment (Justkant) [reply apps/ledger-live-desktop/src/internal/transportHandler.ts:34]: After some digging I think it's better to keep withDevice because without it I would redo a lot of the logic found inside this function
